testsuite/classlibrary/TestArray.sc

```diff
@@ -162,6 +162,29 @@ TestArray : UnitTest {
 
 	} // End test_arraystats
 
+	test_flop_noGC_regression {
+		var a, b, arraySize;
+
+
+		arraySize = 32; // this causes flop to fail
+
+		1000.do{ |i|
+			var indices;
+			indices = (0..arraySize);
+
+			// run identical processing on idx:
+			a = [indices].flop.collect{ arg lm; lm[0] }; // << .lm called in the loop
+			b = [indices].flop.collect{ arg lm; lm[0] }; // << .lm called in the loop
+
+			// are results equal? >> NO
+			this.assertEquals(a, b, "flop: identical flops should match", onFailure: {
+				format("\nWARNING: mismatch on iter: %", i).postln;
+				a.join(" ").postln;
+				b.join(" ").postln;
+			});
+		}
+	}
+
```

Comment:
i wonder if this should really go in this file.. it is not testing `Array` so much as gc behavior. but we also don't have a convention yet (under the new unit testing guidelines) for naming files that don't test a class. my first thought would be to place it in a file/class `Test_GC`, underscore to indicate what follows is not a class name. not sure.

Reply:
Let's leave it here for now. It's actually a bit of a weird test, as it's really targeting a particular regression. (Could happen to catch other things, but...)

What I think we should really do is make a functionality test for every primitive that does object creation (and maybe just most or all primitives), and then make some specific very targeted tests that check bits of GC functionality, perhaps with test primitives for that purpose. But I think that's more for something like #3384.

Reply:
> Let's leave it here for now. 

ok, could you leave a FIXME/TODO so the technical debt is noted then?

> It's actually a bit of a weird test, as it's really targeting a particular regression

we have lots of these tests, i like them a lot actually, they're kind of like battle scars :)

> What I think we should really do is make a functionality test for every primitive that does object creation (and maybe just most or all primitives), and then make some specific very targeted tests that check bits of GC functionality, perhaps with test primitives for that purpose.

sgtm, some of these tests should be at the C++ level if possible

Reply:
also if you are going to leave it here i have some code style comments:

- tighten it up by declaring a and b inside the do loop (unless that change eliminates the failure prior to this change? that seems unlikely though) and removing extra empty lines (in general no need to have 2 empty lines in a row)
- `1000.do {|i|` to conform to our style guide
- `.collect { |lm| lm[0] }` to conform to our style guide
- since there is only 1 semantic thing tested here (that gc "works" in a non-regressing way) there should only be 1 assert. details of any failures can be added in the `details:` parameter of `assertEquals`. passing an `onFailure:` argument will cause test running to end immediately on a failure, which is probably not what you want based on the message. i would accumulate details on any failure during the loop (in string format) and then have one assert outside the loop at the end of the test, with the information you're posting in `onFailure` instead passed as `details:`.

Reply:
> * tighten it up by declaring a and b inside the do loop (unless that change eliminates the failure prior to this change? that seems unlikely though) and removing extra empty lines (in general no need to have 2 empty lines in a row)

Sorry, but no way. It is not totally clear to me that this won't affect the test, and given how quirky GC issues can be and how particular this is, if we're doing this at all I'd like to be sure we're testing what we thought we were. Providing you can prove conclusively that that won't make a difference, and the outer declaration really bugs you, go ahead, but for me that's a step too far into time wasting finickiness, sorry.

> * i would accumulate details on any failure during the loop (in string format) and then have one assert outside the loop at the end of the test, with the information you're posting in `onFailure` instead passed as `details:`.

I considered that, but the number of fails isn't relevant, so the end on first failure is actually intentional. The number of fails will vary a bit based on the state of the GC, but tells you nothing of use. Any fail means you've got the problem (or at least a problem).

The rest I'll do shortly.

Reply:
i'm not trying to waste your time.

Reply:
there cannot be 1000 asserts made by this test. it will flood test runner output. please change it so only 1 assert is made. and please use pipes instead of arg on lines 175 and 176 as i noted in the earlier review.

Reply:
> i'm not trying to waste your time.

Sorry. Written in haste. I meant it would be a waste of my time, and frankly yours, to go to the trouble of proving that the test was not affected by this, not that your request was intended to waste my time.

> please use pipes instead of arg on lines 175 and 176 as i noted in the earlier review.

Ah okay. This is a perfect example of why it would be easier just to make those small changes. You weren't explicit about what you wanted changed, and as I didn't compare it character by character, I noticed the bracket and changed that.

Reply:
github used to have a "suggest diff" feature that i hope they bring out of beta, because it's exactly what you're looking for. but i will not make commits directly to someone else's branch unless they request or OK that specific diff, personal preference.